File: backend/app/models/classification.py
"""Structured diet, ingredient flags, and allergen declarations."""

from enum import Enum

from pydantic import BaseModel, ConfigDict


class DietType(str, Enum):
    vegetarian = "vegetarian"
    vegan = "vegan"
    pescetarian = "pescetarian"
    omnivore = "omnivore"


class Allergen(str, Enum):
    gluten = "gluten"
    milk = "milk"
    egg = "egg"
    soy = "soy"
    nuts = "nuts"
    peanuts = "peanuts"
    sesame = "sesame"
    fish = "fish"
    crustaceans = "crustaceans"
    celery = "celery"
    mustard = "mustard"
    sulfites = "sulfites"
    lupin = "lupin"
    molluscs = "molluscs"


ALLERGEN_CODES: frozenset[str] = frozenset(a.value for a in Allergen)

# German display labels. Must stay in sync with
# frontend/src/constants/classification.ts ALLERGEN_LABELS_DE (parity-checked
# by test_allergen_labels_match_frontend). Used to let a search for an
# allergen-group word (e.g. "fisch") also match items via their already
# audited `allergens` list (see scripts/derive_allergens.py) — not a new
# keyword list, just reusing that one.
ALLERGEN_LABELS_DE: dict[Allergen, str] = {
    Allergen.gluten: "Gluten",
    Allergen.milk: "Milch",
    Allergen.egg: "Ei",
    Allergen.soy: "Soja",
    Allergen.nuts: "Schalenfrüchte",
    Allergen.peanuts: "Erdnüsse",
    Allergen.sesame: "Sesam",
    Allergen.fish: "Fisch",
    Allergen.crustaceans: "Krebstiere",
    Allergen.celery: "Sellerie",
    Allergen.mustard: "Senf",
    Allergen.sulfites: "Sulfite",
    Allergen.lupin: "Lupinen",
    Allergen.molluscs: "Weichtiere",
}


class IngredientFlags(BaseModel):
    """Declared ingredients / composition flags (boolean)."""

    model_config = ConfigDict(extra="forbid")

    contains_meat: bool = False
    contains_pork: bool = False
    contains_poultry: bool = False
    contains_beef: bool = False
    contains_fish: bool = False
    contains_shellfish: bool = False
    contains_dairy: bool = False
    contains_egg: bool = False
    contains_honey: bool = False
    contains_alcohol: bool = False
    contains_gelatin: bool = False

File: frontend/src/constants/classification.ts
export const DIET_TYPES = ["vegetarian", "vegan", "pescetarian", "omnivore"] as const;
export type DietType = (typeof DIET_TYPES)[number];

export const DIET_LABELS_DE: Record<DietType, string> = {
  vegetarian: "Vegetarisch",
  vegan: "Vegan",
  pescetarian: "Pescetarisch",
  omnivore: "Omnivor",
};

export const ALLERGENS = [
  "gluten",
  "milk",
  "egg",
  "soy",
  "nuts",
  "peanuts",
  "sesame",
  "fish",
  "crustaceans",
  "celery",
  "mustard",
  "sulfites",
  "lupin",
  "molluscs",
] as const;
export type AllergenCode = (typeof ALLERGENS)[number];

export const ALLERGEN_LABELS_DE: Record<AllergenCode, string> = {
  gluten: "Gluten",
  milk: "Milch",
  egg: "Ei",
  soy: "Soja",
  nuts: "Schalenfrüchte",
  peanuts: "Erdnüsse",
  sesame: "Sesam",
  fish: "Fisch",
  crustaceans: "Krebstiere",
  celery: "Sellerie",
  mustard: "Senf",
  sulfites: "Sulfite",
  lupin: "Lupinen",
  molluscs: "Weichtiere",
};

export const INGREDIENT_FLAG_KEYS = [
  "contains_meat",
  "contains_pork",
  "contains_poultry",
  "contains_beef",
  "contains_fish",
  "contains_shellfish",
  "contains_dairy",
  "contains_egg",
  "contains_honey",
  "contains_alcohol",
  "contains_gelatin",
] as const;
export type IngredientFlagKey = (typeof INGREDIENT_FLAG_KEYS)[number];

export const INGREDIENT_FLAG_LABELS_DE: Record<IngredientFlagKey, string> = {
  contains_meat: "Fleisch",
  contains_pork: "Schweinefleisch",
  contains_poultry: "Geflügel",
  contains_beef: "Rind",
  contains_fish: "Fisch",
  contains_shellfish: "Schalentiere",
  contains_dairy: "Milchprodukte",
  contains_egg: "Ei",
  contains_honey: "Honig",
  contains_alcohol: "Alkohol",
  contains_gelatin: "Gelatine",
};
